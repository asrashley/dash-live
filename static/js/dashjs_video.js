/* jshint esversion: 5, varstmt: false */
/* globals $, console, dashjs */

(function(){
    'use strict';
    var dashlive = window.dashlive;

    if (window.dashlive === undefined) {
        window.dashlive = dashlive = {};
    }

    dashlive.setupDashjsPlayer = function(vid, source) {
	var $vid, player;

	$vid = $(vid);
	$vid.attr('controls', '');
        player = dashjs.MediaPlayer().create();
        player.initialize(vid, source, true);
        $(vid).on('canplay',function(){
	    console.log('start playback');
            vid.play();
        });
    };

    if (dashlive.setupVideoPlayer === undefined) {
        dashlive.setupVideoPlayer = dashlive.setupDashjsPlayer;
    }
})(window);
