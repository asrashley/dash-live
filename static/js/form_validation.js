/**
 * Create an object with all the values from an HTML form
 * @param form {HTMLFormElement} the form to extract data from
 * @param exclude {Set} optional set of keys to exclude
 */
export function serializeForm(form, exclude) {
    const formData = new FormData(form);
    const data = {};
    for (const [key, value] of formData) {
        if (exclude === undefined || !exclude.has(key)) {
            data[key] = value;
        }
    }
    return data;
}

export async function validateFormValues(form_id) {
  const form = document.getElementById(form_id);
  if (form === null) {
    throw new Error('Failed to find form');
  }
  const url = form.getAttribute('data-validate-url');
  if (!url) {
    throw new Error('Failed to find URL for form validation');
  }
  const data = serializeForm(form);
  form.classList.add('needs-validation');
  form.classList.remove('was-validated');

  const fetchResult = await fetch(url, {
    body: JSON.stringify(data),
    cache: 'no-cache',
    credentials: 'same-origin',
    headers: {
      "Content-Type": "application/json",
    },
    method: 'POST',
    mode: 'same-origin',
  });
  if (!fetchResult.ok) {
    throw new Error(`Failed to check form validitity: ${ fetchResult.status }`);
  }
  const jsonResult = await fetchResult.json();
  const { csrf_token, errors = {} } = jsonResult ?? {};
  if (csrf_token) {
    form.csrf_token.value = csrf_token;
  }

  form.querySelectorAll('input').forEach((elt) => {
    if (elt.getAttribute('type') == 'hidden') {
      return;
    }
    const name = elt.getAttribute('name');
    const parent = elt.parentElement;
    if (name === null || parent === null) {
      return;
    }
    const feedback = parent.querySelector('.invalid-feedback');
    const err = errors[name] || '';
    if (err === '') {
      elt.classList.add('is-valid');
      elt.classList.remove('is-invalid');
    } else {
      elt.classList.remove('is-valid');
      elt.classList.add('is-invalid');
    }
    if (feedback !== null) {
      feedback.innerText = err;
    }
    elt.setCustomValidity(err);
  });
  form.classList.remove('needs-validation');
  form.classList.add('was-validated');
  return {
    form,
    ...jsonResult,
  };
}

export function setupValidation(form_id, options = {}) {
  const form = document.getElementById(form_id);
  const { cleanupOnSubmit = true } = options;

  if (form === null) {
    throw new Error(`Failed to find form "${ form_id }"`);
  }
  const controller = new AbortController();
  const { signal } = controller;

  const checkFormValues = () => {
    validateFormValues(form_id).catch(err => console.error(err));
  };

  for (let i = 0; i < form.elements.length; ++i) {
    const elt = form.elements[i];
    if (elt.nodeName === 'INPUT') {
      elt.addEventListener('change', checkFormValues, { signal, });
      elt.addEventListener('invalid', (ev) => {
        ev.target.classList.remove('is-valid');
        ev.target.classList.add('is-invalid');
      }, { signal, });
      elt.addEventListener('valid', (ev) => {
        ev.target.classList.add('is-valid');
        ev.target.classList.remove('is-invalid');
      }, { signal, });
    }
  }
  if (cleanupOnSubmit) {
    const submitHandler = () => {
      controller.abort();
      form.removeEventListener('submit', submitHandler);
    };
    form.addEventListener('submit', submitHandler);
  }
  return {form, controller, signal,};
}
