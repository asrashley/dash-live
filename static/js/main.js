$(document).ready(function(){
  'use strict';
  var pageState = {
    title: document.title,
    url: location.href,
    baseurl: location.href.split('#')[0],
  };
  var cursor = { row:0, col:0, stream:0 };

  function cycleOption(cell) {
    var sel, options, cur, i;

    sel = cell.find('select');
    options = cell.find('option');
    cur = sel.val();
    for (i=0; i < options.length; ++i) {
      if (options[i].value == cur) {
        sel.val(options[(i+1)%options.length].value);
        return;
      }
    }
  }

  function moveCursor(direction) {
    var xpos, ypos, rows, cols;

    xpos = cursor.col;
    ypos = cursor.row;
    rows = $('#buttons tbody tr');
    cols = $('#buttons thead th');

    switch(direction) {
    case 'up':
      ypos -= 1;
      if (ypos < 0) {
        ypos = rows.length - 1;
      }
      break;
    case 'down':
      ypos = (ypos+1) % rows.length;
      break;
    case 'right':
          if (cursor.row < (cursor.numRows-1)) {
                xpos = (xpos+1) % cols.length;
          } else {
                cursor.stream = (cursor.stream+1) % pageState.streams.length;
                $('#play-button select').val(pageState.streams[cursor.stream].directory);
                updateManifestURL();
          }
      break;
    case 'left':
          if (cursor.row < (cursor.numRows-1)) {
                xpos -= 1;
                if (xpos < 0) {
          xpos = cols.length - 1;
                }
          } else {
                cursor.stream--;
                if (cursor.stream<0) {
                  cursor.stream = pageState.streams.length - 1;
                  $('#play-button select').val(pageState.streams[cursor.stream].directory);
                  updateManifestURL();
                }
          }
      break;
    }
    $('#buttons .cursor').removeClass('cursor');
    cols = $(rows[ypos]).find('td');
    $(cols[xpos % cols.length]).addClass('cursor');
    $('#buttons .active').removeClass('active');
      if (ypos == (rows.length)-1) {
          $('#buttons tbody tr:last-child').addClass('active');
      } else {
          $('#buttons tbody .manifest:nth-child('+(xpos+1)+'), #buttons tbody .option:nth-child('+(xpos+1)+')').addClass('active');
      }
      cursor.row = ypos;
      cursor.col = xpos;
  }

  function buildCGI() {
    var params;

    params = {
        mode: "mode=vod"
    };
    $('#buttons tbody .option select').each(function(idx, sel) {
	var defaultVal, val;
	val = $(sel).val();
	defaultVal = $(sel).data('default');
	if (sel.name === 'mode') {
            params[sel.name] = val;
	}
	else if (val && !val.endsWith('=none') && val != defaultVal) {
            params[sel.name] = val;
	}
    });
    if (params.drmloc) {
      if (params.drm && params.drm!=="drm=none") {
        params.drm += '-' + params.drmloc.split('=')[1];
      }
      delete params.drmloc;
    }
    return params;
  }

  function calculateManifestURL(params) {
    var manifest, url;

    manifest = $('#buttons tbody .manifest.selected');
    if (manifest.length == 0) {
      return null;
    }
    url = manifest.data("uri").replace('{directory}', pageState.streams[cursor.stream].directory);
    url = url.replace("{mode}", params.mode.slice(5));
    delete params.mode;
    return url;
  }

  function updateManifestURL() {
    var manifest, dest, url, params, params_str;

    dest = $('#dashurl');
    params = buildCGI();
    url = calculateManifestURL(params);
    if (url === null) {
      dest.text('');
      dest.attr('href','');
      $('#play-button').addClass('disabled');
      return;
    }
    params_str = Object.values(params).join('&');
    if (params_str) {
      url += '?' + params_str;
    }
    dest.text(url);
    dest.attr('href', document.location.origin + url);
    dest = $('#view-manifest');
    dest.attr('href', url.replace('/dash/', '/view/'));
    dest.removeClass('disabled');
    $('#play-button').removeClass('disabled');
    if(window.history && typeof(history.pushState)==="function") {
	params.mpd = 'mpd=' + $('#buttons tbody .manifest.selected').data("filename");
	url = pageState.baseurl + '#' + Object.values(params).join('&');
	history.replaceState(pageState, $('#buttons tbody .manifest.selected').text(), url);
    }
  }

  function playVideo() {
    var params, url;

    if (!$('#dashurl').attr("href")) {
      return;
    }
    params = buildCGI();
    url = calculateManifestURL(params);
    if (url === null) {
      return;
    }
    url = url.replace(/^\/dash/, '/play');
    url = url.replace(/\.mpd$/, '/index.html');
    delete params.mode;
    url += '?' + Object.values(params).join('&');
    document.location = url;
  }

  $('body').on('keydown', function(ev){
    var row, active, params, p;

    if(ev.keyCode>=48 && ev.keyCode<=57){
      row = $('#row'+String.fromCharCode(ev.keyCode));
      if (row.length===0) {
        return;
      }
      active = row.find(".active");
      if (active.length===0) {
        return;
      }
      if (active.hasClass("manifest")) {
        $('#buttons .selected').removeClass('selected');
        active.addClass("selected");
        params = active.data('params');
        for (p in params) {
          $('#buttons select[name="'+p+'"]').val(p+'='+params[p]);
        }
        moveCursor('right');
      } else {
        cycleOption(active);
      }
      updateManifestURL();
      return;
    }
    switch(ev.keyCode) {
    case 13: /* enter */
      active = $('#row'+cursor.row+' td:nth-child('+(cursor.col+1)+')');
      if (active.length===0) {
        /* cursor is on the "play video" row */
        playVideo();
        return;
      }
      if (cursor.col==0) {
        $('#buttons .selected').removeClass('selected');
        active.addClass("selected");
        params = active.data('params');
        for (p in params) {
          $('#buttons select[name="'+p+'"]').val(p+'='+params[p]);
        }
        updateManifestURL();
        moveCursor('right');
      } else {
        cycleOption(active);
      }
      break;
    case 80: /* 'p' */
      /* falls through */
    case 415: /* PLAY */
      playVideo();
      break;
    case 37: /* left */
          moveCursor('left');
      break;
    case 38: /* up */
      moveCursor('up');
      break;
    case 39: /* right */
          moveCursor('right');
      break;
    case 40: /* down */
      moveCursor('down');
      break;
    default:
      console.log(ev.keyCode);
      break;
    }
  });

  function init() {
    var params, i, p;

    $('a.manifest-title').on('click', function(ev) {
      var cell = ev.target;
      ev.preventDefault();
      $('#buttons .selected').removeClass('selected');
      while (cell.nodeName!="TD") {
        cell = cell.parentNode;
      }
      $(cell).addClass("selected");
      updateManifestURL();
      return false;
    });
    $('th.manifest, th.option').on('click', function(ev) {
      if ($(ev.target).hasClass('active')) {
        return;
      }
      toggleActiveRow();
    });
    $('#buttons tbody .option select').on('change', function(ev) {
      updateManifestURL();
    });
    $('#play-button select').on('change', function(ev) {
          var i, val = $(ev.target).val();
          for(i=0; i<pageState.streams.length; ++i) {
                if (pageState.streams[i].directory == val) {
                  cursor.stream = i;
                }
          }
      updateManifestURL();
    });
    $('.play-icon').on('click', playVideo);

    if (typeof(Object.values)==="undefined") {
      Object.values = function(obj) {
        var k, rv = [];
        for (k in obj) {
          if (obj.hasOwnProperty(k)) {
            rv.push(obj[k]);
          }
        }
        return rv;
      }
      Object.keys = function(obj) {
        var k, rv = [];
        for (k in obj) {
          if (obj.hasOwnProperty(k)) {
            rv.push(k);
          }
        }
        return rv;
      }
    }

    cursor.numRows = $('#buttons tbody tr').length;
    cursor.numCols = $('#buttons tbody tr:first-child td').length;
    pageState.streams = [];
    $('#play-button option').each(function(idx, opt) {
        opt = $(opt);
        pageState.streams.push({title: opt.text(), directory: opt.val()});
        if (opt.val() == $(".play-select").val()) {
            cursor.stream = idx;
        }
    });
    if (/#/.test(document.location.href)) {
      params = document.location.href.split('#')[1];
      params = params.split('&');
      for(i=0; i<params.length; ++i) {
        p = params[i].split('=');
        if (p[0]=='mpd') {
          $('#buttons .selected').removeClass('selected');
          /* console.log('#buttons .manifest[data-filename="'+p[1]+'"]');
          $('#buttons .manifest[data-filename="'+p[1]+'"]').css('background-color','red'); */
          $('#buttons .manifest[data-filename="'+p[1]+'"]').addClass('selected');
        } else {
          /* $('#buttons select[name="'+p[0]+'"]').css('background-color','red'); */
          $('#buttons select[name="'+p[0]+'"]').val(params[i]);
        }
      }
      updateManifestURL();
    }
    moveCursor('');
    document.body.className += ' '+navigator.platform.toLowerCase();
  }
  init();
});
