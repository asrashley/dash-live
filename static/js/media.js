$(document).ready(function () {
    'use strict';
    function addClickListeners() {
        $('#keys .add-key').on('click', addKey);
        $('#keys .delete-key').on('click', deleteKey);
        $('#streams .add-stream').on('click', addStream);
        $('#streams .delete-stream').on('click', deleteStream);
        $('#edit-model .delete-stream').on('click', deleteStream);
        $('#media-files .delete-file').on('click', deleteFile);
        $('#media-files .btn-index').on('click', indexFile);
        $("#upload-form .submit").on('click', uploadFile);
        $('#dialog-box .btn-close').on('click', closeDialog);
    }

    function removeClickListeners() {
        $('#keys .add-key').off('click', addKey);
        $('#keys .delete-key').off('click', deleteKey);
        $('#streams .add-stream').off('click', addStream);
        $('#streams .delete-stream').off('click', deleteStream);
        $('#edit-model .delete-stream').off('click', deleteStream);
        $('#media-files .delete-file').off('click', deleteFile);
        $('#media-files .btn-index').off('click', indexFile);
        $("#upload-form .submit").off('click', uploadFile);
        $('#dialog-box .btn-close').off('click', closeDialog);
    }

    function reloadContent() {
        $.ajax({
            url: document.location + '?fragment=1',
            method: 'GET',
        }).done(function (result) {
            removeClickListeners();
            $('#media-management').parent().html(result);
            addClickListeners();
        });
    }
    function addKey(ev) {
        var kid, key, csrf;
        var $row = $(ev.target).parents('tr');
        kid = $row.find('input[name="kid"]').val();
        key = $row.find('input[name="key"]').val();
        csrf = $('#keys').data('csrf');
        $row.find('.btn').attr("disabled", true);
        $('#keys .error').text('');
        $.ajax({
            url: '/key',
            method: 'PUT',
            data: {
                "kid": kid,
                "key": key,
                "csrf_token": csrf
            },
            dataType: 'json',
        }).done(function (result) {
            var newRow;
            if (result.error) {
                $('#keys .error').text(result.error);
                return;
            }
            if (result.csrf_token) {
                $('#keys').data('csrf', result.csrf_token);
            }
            reloadContent();
        });
    }

    function deleteKey(ev) {
        var kid, key, csrf;
        var $row = $(ev.target).parents('tr');
        kid = $row.find('.kid').text();
        csrf = $('#keys').data('csrf');
        if (!kid) {
            return;
        }
        $('#keys .error').text('');
        $.ajax({
            url: '/key/' + kid + "?csrf_token=" + csrf,
            method: 'DELETE',
            dataType: 'json',
        }).done(function (result) {
            if (result.error) {
                $('#keys .error').text(result.error);
            } else {
                $row.remove();
            }
            if (result.csrf) {
                $('#keys').data('csrf', result.csrf);
            }
        }).fail(function (jqXhr, status) {
            $('#keys .error').text(status);
        });
    }

    function addStream(ev) {
        var data;
        var $row = $(ev.target).parents('tr');
        data = {
            'title': $row.find('input[name="title"]').val(),
            'directory': $row.find('input[name="directory"]').val(),
            'marlin_la_url': $row.find('input[name="marlin_la_url"]').val(),
            'playready_la_url': $row.find('input[name="playready_la_url"]').val(),
            'csrf_token': $('#streams').data('csrf')
        };
        $row.find('.btn').attr("disabled", true);
        $('#streams .error').text('');
        $.ajax({
            url: '/stream',
            method: 'PUT',
            contentType: "application/json; charset=utf-8",
            data: JSON.stringify(data),
            dataType: 'json'
        }).done(function (result) {
            var newRow;
            if (result.error) {
                $('#streams .error').text(result.error);
                return;
            }
            if (result.csrf) {
                $('#streams').data('csrf', result.csrf);
            }
            reloadContent();
        });
    }

    function deleteStream(ev) {
        var dialog, id, title, $row, csrf, input;

        function onClick(btnEv) {
            var cmd;

            btnEv.preventDefault();
            closeDialog();
            cmd = $(btnEv.target).data("cmd");
            if (cmd === "yes") {
                confirmDeleteStream($row, id, csrf);
            }
            return false;
        }

        ev.preventDefault();

        id = $(ev.target).data("id");
        if (!id) {
            return false;
        }
        csrf = $('#streams').data('csrf');
        if (!csrf) {
            input = $('#edit-model input[name="csrf_token"]');
            if (!input) {
                return false;
            }
            csrf = input[0].value;
        }
        $row = $(ev.target).parents('tr');
        if ($row.length) {
            title = $row.find('.title').text();
            if (title === "" || title === undefined) {
                title = $row.find('.directory').text();
            }
        } else {
            title = ev.target.innerText.slice(7, -1);
        }
        dialog = $('#dialog-box');
        dialog.find(".modal-body").html(
            '<div><h3>Delete stream &quot;' + title + '&quot; ?</h3>' +
                '<p>This will also delete all of the media files for this stream</p></div>');
        dialog.find(".modal-footer").html(
            '<div>' +
            '<button class="btn btn-danger" style="margin:1em" data-cmd="yes">Delete Stream</button>' +
            '<button class="btn btn-secondary" data-cmd="no">Cancel</button>' +
            '</div>');
        dialog.find(".modal-footer .btn").one('click', onClick);
        showDialog();
        return false;
    }

    function confirmDeleteStream($row, id, csrf) {
        if (!id) {
            return;
        }
        $('#streams .error').text('');
        $.ajax({
            accepts: 'application/json',
            url: `/stream/${id}`,
            method: 'DELETE',
            dataType: 'json',
            data: {
                csrf_token: csrf,
                ajax: 1,
            },
        }).done(function (result) {
            if (result.error) {
                $('#streams .error').text(result.error);
            } else {
                window.location = $('#media-management').data('back');
            }
        }).fail(function (jqXhr, status) {
            $('#streams .error').text(status);
        });
    }

    function indexFile(ev) {
        var dialog, blobId, filename, csrf;
        var $row = $(ev.target).parents('tr');
        var $btn = $(ev.target);
        blobId = $btn.data('key');
        filename = $row.find('.filename').text()
        if (!blobId) {
            return;
        }
        csrf = $('#media-files').data('csrf');
        dialog = $('#dialog-box');
        dialog.find(".modal-body").html('<p>Indexing ' + filename +
            '</p><div class="error"></div>');
        showDialog();
        $.ajax({
            url: '/media/index/' + blobId + '?csrf_token=' + csrf,
            method: 'GET',
            dataType: 'json',
        }).done(function (result) {
            var i;
            if (result.error) {
                dialog.find('.modal-body .error').text(result.error);
            } else {
                dialog.find(".modal-body").html('<p>Indexing ' +
                    filename + ' complete</p>');
                if (result.representation) {
                    $row.find('td.codec').text(result.representation.codecs);
                    if (result.representation.encrypted) {
                        $row.find('td.encrypted').html('<span class="bool-yes ">&check;</span>');
                        $row.find('td.kid').html("");
                        for (i = 0; i < result.representation.kids.length; ++i) {
                            $row.find('td.kid').append(
                                '<p>' +
                                encodeURIComponent(result.representation.kids[i]) +
                                '</p>');
                        }
                    } else {
                        $row.find('td.encrypted').html('<span class="bool-no ">&cross;</span>');
                    }
                    $row.find('.btn-index').addClass('btn-info').removeClass('btn-warning').text('Re-index');
                    window.setTimeout(closeDialog, 750);
                }
            }
            if (result.csrf) {
                $('#media-files').data('csrf', result.csrf);
            }
        }).fail(function (e) {
            var err = dialog.find('.modal-body .error');
            if (e.statusText) {
                err.text(e.status + ' ' + e.statusText);
            } else if (e.responseText) {
                err.text(e.responseText);
            } else {
                err.text(JSON.stringify(e));
            }
        });
    }

    function deleteFile(ev) {
        var href, csrf;
        var $row = $(ev.target).parents('tr');
        var $btn = $(ev.target);
	ev.preventDefault();
        href = $btn.attr('href');
        csrf = $('#media-files').data('csrf');
        $('#media .error').text('');
        $.ajax({
            url: href + '?csrf_token=' + csrf,
            method: 'DELETE',
            dataType: 'json',
        }).done(function (result) {
            if (result.error) {
                $('#media .error').text(result.error);
            }
            if (result.csrf) {
                $('#media-files').data('csrf', result.csrf);
            }
	    reloadContent();
        }).fail(function (jqXhr, status) {
            $('#media .error').text(status);
        });
    }
    function uploadFile(ev) {
        var form, data, dialog, filename;
        ev.preventDefault();
        form = $("#upload-form");
        filename = form.find('input[name="file"]').val();
        if (filename === "") {
            alert("No file selected");
            return;
        }
        form.find('input[name="ajax"]').val("1");
        data = new FormData(form[0]);
        $("#upload-form .submit").prop("disabled", true);
        dialog = $('#dialog-box');
        dialog.find(".modal-body").html('<p>Uploading ' +
            encodeURIComponent(filename) +
            '</p><div class="error"></div>');
        showDialog();
        $.ajax({
            url: form.attr("action"),
            data: data,
            type: "POST",
            enctype: 'multipart/form-data',
            processData: false,
            contentType: false,
            timeout: 600000,
            cache: false
        }).done(function (data) {
            var err, htm;

            $("#btnSubmit").prop("disabled", false);
            if (data.error) {
                err = dialog.find('.modal-body .error');
                err.text(data.error);
                return;
            }
            dialog.find(".modal-body").html('<p>Finished uploading ' +
                filename +
                '<span class="bool-yes ">&check;</span>');
            if (data.upload_url) {
                $('#upload-form').attr('action', data.upload_url);
            }
            if (data.csrf_token) {
                $('#upload-form input[name="csrf_token"]').val(data.csrf_token);
            }
	    closeDialog();
            reloadContent();
        }).fail(function (e) {
            var err = dialog.find('.modal-body .error');
            if (e.responseJSON) {
                err.text(e.responseJSON.error);
            }
            else if (e.statusText) {
                err.text(e.status + ' ' + e.statusText);
            } else if (e.responseText) {
                err.text(e.responseText);
            } else {
                err.text(JSON.stringify(e));
            }
            console.error(e);
	    closeDialog();
        });
        return false;
    }

    function showDialog() {
        var dialog = $('#dialog-box');
        dialog.addClass("dialog-active show");
        dialog.css({ display: "block" });
        $('.modal-backdrop').addClass('show');
        $('.modal-backdrop').removeClass('hidden');
        $('body').addClass('modal-open');
    }

    function closeDialog() {
        var dialog = $('#dialog-box');
        dialog.removeClass("dialog-active").removeClass("show");
        dialog.css("display", "");
        $(document.body).removeClass("modal-open");
        $('.modal-backdrop').addClass('hidden');
        $('.modal-backdrop').removeClass("show");
    }

    addClickListeners();
});
