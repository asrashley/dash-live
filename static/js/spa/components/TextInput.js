import { html } from 'htm/preact';

export function TextInput({name, error, value, onInput, required}) {
  const className = `form-control ${error ? "is-invalid" : "is-valid"}`;
  return html`
    <input class="${className}" id="field-${name}" name="${name}"
      type="text" value=${value} onInput=${onInput}
      required=${required} />`;
}
