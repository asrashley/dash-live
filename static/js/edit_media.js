/* globals console */
import $ from '/libs/jquery.js';
import { enableTooltips } from './tooltips.js';

async function checkFormValues() {
    const form = document.getElementById('edit_media');
    if (form === null) {
        console.error('Failed to find edit media form');
        return;
    }
    const url = $(form).data('validate-url');
    if (!url) {
        console.error('Failed to find URL for form validation');
        return;
    }
    const data = Object.fromEntries(
        $(form).serializeArray().
            filter(item => item.name !== 'csrf_token').
            map(item => [item.name, item.value,]));
    $('form').addClass('needs-validation').removeClass('was-validated');
    const result = await fetch(url, {
        body: JSON.stringify(data),
        cache: 'no-cache',
        credentials: 'same-origin',
        headers: {
            "Content-Type": "application/json",
        },
        method: 'POST',
        mode: 'same-origin',
    });
    if (!result.ok) {
        console.error(`Failed to check form validitity: ${ result.status }`);
        return;
    }
    const { errors } = await result.json();
    $('#edit_media input').each((_index, elt) => {
        const $elt = $(elt);
        const name = $elt.attr('name');
        const parent = $elt.parent();
        const feedback = parent.find('.invalid-feedback');
        const err = errors[name] || '';
        if (err === '') {
            $elt.addClass('is-valid').removeClass('is-invalid');
        } else {
            $elt.removeClass('is-valid').addClass('is-invalid');
        }
        feedback.text(err);
        elt.setCustomValidity(err);
    });
    $('form').removeClass('needs-validation').addClass('was-validated');
    const submit = $('#edit_media button[type="submit"]');
    if (form.checkValidity()) {
        submit.attr('disabled', false);
    } else {
        submit.attr('disabled', 'disabled');
    }
}

function setupValidation() {
    $('#edit_media input').on('change', checkFormValues);

    $('#edit_media input').on('invalid', (ev) => {
        $(ev.target).removeClass('is-valid').addClass('is-invalid');
    });

    $('#edit_media input').on('valid', (ev) => {
        $(ev.target).addClass('is-valid').removeClass('is-invalid');
    });
}

enableTooltips();
setupValidation();
