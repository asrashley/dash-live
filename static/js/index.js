import $ from 'jquery';
import { defaultOptions } from '/libs/options.js';

const pageState = {
    title: document.title,
    url: location.href,
    baseurl: location.href.split('#')[0],
};

function buildCGI() {
  const params = {};
  const items = $('#id_main_form').serializeArray();
  items.forEach((item) => {
    if (item.value === "") {
      return;
    }
    if (/advanced|csrf_token|manifest|mode|model|mpd|stream/.test(item.name)) {
      return;
    }
    const dflt = defaultOptions[item.name];
    if (item.value !== `${dflt}`) {
      if (params[item.name] !== undefined) {
        if (typeof params[item.name] !== 'object') {
          params[item.name] = [params[item.name], item.value,];
        } else {
          params[item.name].push(item.value);
        }
      } else {
        params[item.name] = item.value;
      }
    }
  });
  const drms = [];
  ['clearkey', 'marlin', 'playready',].forEach(name => {
    if (params[`${name}__enabled`] === 'on') {
      const drmloc = `${name}_drmloc`;
      if (params[drmloc] === undefined || params[drmloc] === "") {
        drms.push(name);
      } else {
        drms.push(`${name}-${params[drmloc]}`);
      }
      delete params[drmloc];
      delete params[`drm_${name}`];
    }
  });
  if (drms.length) {
    params.drm = `${drms.join(',')}`;
  }
  const rv = {};
  for (const [key, value] of Object.entries(params)) {
    rv[key] = `${key}=${value}`;
  }
  return rv;
}

function getSelectedMode() {
  const options = $('input[name="mode"]');
  for(let i=0; i < options.length; ++i) {
    if (options[i].checked) {
      return {
        value: options[i].value,
        title: options[i].labels[0].innerText,
      };
    }
  }
  return null;
}

function getSelectedStream() {
  const options = $('input[name="stream"]');
  for(let i=0; i < options.length; ++i) {
    if (options[i].checked) {
      return {
        value: options[i].value,
        title: options[i].labels[0].innerText,
      };
    }
  }
  return null;
}

function calculateManifestURL() {
  const stream = getSelectedStream();
  if (!stream) {
    return '';
  }
  const manifest = $('#model-manifest').val();
  const mode = getSelectedMode();
  const isMpsStream = /^mps\./.test(stream.value);
  const template = $('#with-modules').data(
    isMpsStream ? 'mps-url-template': 'stream-url-template');
  const name = stream.value.split('.')[1];

  let url = template.replace('{manifest}', manifest);
  url = url.replace('{directory}', name);
  url = url.replace('{stream}', name);
  url = url.replace("{mode}", mode.value);
  return url;
}

function updateManifestURL() {
  const stream = getSelectedStream();
  const params = buildCGI();
  const mpdUrl = calculateManifestURL(params);
  let dest = $('a.selected-stream');
  if (mpdUrl === null) {
    dest.text('');
    dest.attr('href','');
    $('.play-button').addClass('disabled');
    $('#dashurl').text('');
    $('.view-manifest').addClass('disabled');
    return;
  }
  let url = mpdUrl.replace(/^\/dash/, '/play');
  url = url.replace(/^\/mps/, '/play/mps');
  url = url.replace(/\.mpd$/, '/index.html');
  const params_str = Object.values(params).join('&');
  if (params_str) {
    url += `?${params_str}`;
  }
  $('a.selected-stream .title').text(`Play ${stream.title}`);
  dest.attr('href', document.location.origin + url);
  $('.play-button').removeClass('disabled');
  $('#with-modules .view-manifest').removeClass('disabled');

  dest = $('#dashurl');
  if (params_str) {
    dest.text(`${mpdUrl}?${params_str}`);
    dest.attr('href', `${document.location.origin}${mpdUrl}?${params_str}`);
  } else {
    dest.text(mpdUrl);
    dest.attr('href', document.location.origin + mpdUrl);
  }
  dest.removeClass('disabled');
  const viewUrl = dest.attr('href').replace(/\/(dash|mps)\//, '/view/$1/')
  $('.view-manifest').attr('href', viewUrl);
  $('#radio-mode-odvod').attr('disabled', /^mps\./.test(stream.value));

  if(window.history && typeof(history.pushState)==="function") {
    const page_params = Object.values(params);
    const modelManifest = $('#model-manifest').val()
    page_params.push(`manifest=${modelManifest}`);
    url = `${pageState.baseurl}#${page_params.join('&')}`;
    history.replaceState(
      pageState, $('#buttons tbody .manifest.selected').text(), url);
  }
}

function showOrHideAdvancedOptions(enable) {
  if (enable) {
    $('.row.advanced').removeClass('hidden');
    $('.accordion-item.advanced').removeClass('hidden');
  } else {
    $('.row.advanced').addClass('hidden');
    $('.accordion-item.advanced').addClass('hidden');
  }
}

function updateDom() {
  $('#group-0 .accordion-header').html(`
    <span class="accordion-button">
      ${$('#group-0 .accordion-header .accordion-button').text()}
    </span>`);
  const allOptions = $("#show-all-options").detach();
  $('#group-0 .accordion-header').append(allOptions);
  $('.row-field-drm_clearkey').removeClass('hidden advanced');
  $('.row-field-drm_marlin').removeClass('hidden advanced');
  $('.row-field-drm_playready').removeClass('hidden advanced');
  $('#model-group-1').append($('.row-field-drm_clearkey').remove());
  $('#model-group-1').append($('.row-field-drm_marlin').remove());
  $('#model-group-1').append($('.row-field-drm_playready').remove());
  $('#with-modules').css('display', 'block');
  $('#buttons').remove();
}

function init() {
  const defaultFields = ['mode', 'stream', 'manifest', 'player'];
  const form = document.getElementById('id_main_form');

  updateDom();
  $('#advanced-options').on('change', (ev) => {
    showOrHideAdvancedOptions(ev.target.checked);
  });
  $('#id_main_form').on('change', updateManifestURL);

  defaultFields.forEach(name => {
    const value = localStorage.getItem(`dashlive.${name}`);
    if (value !== null) {
      form[name].value = value;
    }
    $(`input[name="${name}"], select[name="${name}"]`).on('change', (ev) => {
      const { value } = ev.target;
      if (value) {
        localStorage.setItem(`dashlive.${name}`, value);
      }
    });
  });
  if (/#/.test(document.location.href)) {
    let params = document.location.href.split('#')[1];
    params = params.split('&');
    for(let i=0; i < params.length; ++i) {
      const p = params[i].split('=');
      $(`#model-${p[0]}`).val(p[1]);
    }
  }
  updateManifestURL();
  showOrHideAdvancedOptions(
    document.getElementById('advanced-options').checked);
}

init();
