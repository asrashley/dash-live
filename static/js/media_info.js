import $ from '/libs/jquery.js';
import { decode } from './prod/codec-string.js';
import { enableTooltips } from './tooltips.js';

function setCodecDescription() {
    const codecDescription = decode($('.codec-string').text());
    const details = codecDescription.split('<br>').filter((line) => line !== "");
    $('.codec-description').html(details.join('<br>'));
}

setCodecDescription();
enableTooltips();
