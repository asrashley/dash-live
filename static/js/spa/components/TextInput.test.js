import { describe, expect, test, vi } from "vitest";
import { html } from "htm/preact";

import { renderWithProviders } from "../../test/renderWithProviders.js";
import { TextInput } from "./TextInput.js";

describe("TextInput", () => {
  const onInput = vi.fn();

  test("matches snapshot", () => {
    const { asFragment, getBySelector } = renderWithProviders(
      html`<${TextInput}
        name="tname"
        onInput=${onInput}
        placeholder="search..."
        required=${true}
      />`
    );
    const inp = getBySelector(".form-control");
    expect(inp.className.trim()).toEqual("form-control is-valid");
    expect(inp.getAttribute("type")).toEqual("text");
    expect(inp.getAttribute("id")).toEqual("field-tname");
    expect(asFragment()).toMatchSnapshot();
  });

  test("text input with error", () => {
    const { getBySelector } = renderWithProviders(
      html`<${TextInput}
        name="tname"
        onInput=${onInput}
        error="input has an error"
      />`
    );
    const inp = getBySelector(".form-control");
    expect(inp.className.trim()).toEqual("form-control is-invalid");
  });
});
