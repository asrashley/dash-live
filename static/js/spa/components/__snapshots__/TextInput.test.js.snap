// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInput > matches snapshot 1`] = `
<DocumentFragment>
  <input
    class="form-control is-valid"
    id="field-tname"
    name="tname"
    placeholder="search..."
    required=""
    type="text"
  />
</DocumentFragment>
`;
